File: 1_arrays/4_kadanes_algorithm.py
# TODO Print actual sub array as follow up question
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        maxi_sum = nums[0]
        curr_sum = nums[0]
        for item in nums[1:]:
            curr_sum = max(item, curr_sum + item)
            maxi_sum = max(maxi_sum, curr_sum)
        return maxi_sum
        

File: 01_arrays/4_kadanes_algorithm.rs
use std::cmp::max;

impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut res = nums[0];
        let mut curr = nums[0];

        for num in &nums[1..] {
            if curr < 0 {
                curr = 0;
            }
            curr += num;
            res = max(res, curr);
        }
        res
    }
}
